### PR TITLE
Optimise the fast path for alloc and dealloc

### PR DESCRIPTION
**Pull request by mjp41**
This seems to have a pretty drastic improvement in speed to heavy allocation benchmarks. It pushes `snmalloc` to have approximately the same performance as `mimalloc` on the `cfrac` benchmark.

### CODE REVIEW

**Comment by mjp41** [comment]
This addresses #58

File: src/mem/sizeclass.h [hunk 1]
```diff
@@ -4,35 +4,31 @@
...
-    return static_cast<uint8_t>(
+    return static_cast<sizeclass_t>(
       bits::to_exp_mant_const<INTERMEDIATE_BITS, MIN_ALLOC_BITS>(size));
   }
 
   constexpr static inline size_t large_sizeclass_to_size(uint8_t large_class)
   {
     return bits::one_at_bit(large_class + SUPERSLAB_BITS);
-  }
+  }
```
Comment (davidchisnall) [inline src/mem/sizeclass.h]: Space.

File: src/mem/sizeclass.h [hunk 2]
```diff
@@ -4,35 +4,31 @@
...
-  constexpr static inline uint8_t size_to_sizeclass_const(size_t size)
+
+  constexpr static inline sizeclass_t size_to_sizeclass_const(size_t size)
   {
     // Don't use sizeclasses that are not a multiple of the alignment.
     // For example, 24 byte allocations can be
     // problematic for some data due to alignment issues.
-    return static_cast<uint8_t>(
+    return static_cast<sizeclass_t>(
       bits::to_exp_mant_const<INTERMEDIATE_BITS, MIN_ALLOC_BITS>(size));
```
Comment (davidchisnall) [inline src/mem/sizeclass.h]: Do we want an assert that this is in range for a `uint8_t`?

File: src/mem/sizeclasstable.h [hunk 3]
```diff
@@ -5,31 +5,49 @@
...
 namespace snmalloc
 {
+  constexpr size_t PTR_BITS = bits::next_pow2_bits_const(sizeof(void*));
+
+  constexpr static size_t sizeclass_lookup_index(size_t s)
+  {
+    // We shift by PTR_BITS as makes code-gen for the table lookup nicer.
+    // We could shift by MIN_ALLOC_BITS, but then there is a slightly more
+    // complex sequence.
+    return (s - 1) >> PTR_BITS;
```
Comment (davidchisnall) [inline src/mem/sizeclasstable.h]: I think I was more confused after reading this comment than before.  I guess the key point is that we can index the table by anything we like that's cheap to compute, and this particular shift makes it cheap, at the expense of a larger table?

File: src/mem/sizeclass.h (hunk 1 above)
Comment (mjp41) [reply src/mem/sizeclass.h]: Fix clangformat to actually run on builds, so this should be fixed if it passes CI now.

**Pull request merged**